sql/src/main/java/io/crate/execution/engine/window/WindowBatchIterator.java

```diff
@@ -0,0 +1,133 @@
+import io.crate.data.BatchIterator;
+import io.crate.data.CollectingRowConsumer;
+import io.crate.data.InMemoryBatchIterator;
+import io.crate.data.MappedForwardingBatchIterator;
+import io.crate.data.Row;
+import io.crate.execution.engine.aggregation.AggregateCollector;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+
+public class WindowBatchIterator extends MappedForwardingBatchIterator<Row, Row> {
+
+    private final BatchIterator<Row> source;
+    private final AggregateCollector collector;
+    private WindowDefinition windowDefinition;
+
+    private Row currentElement;
+
+    private int sourceRowsConsumed;
+    private int rowPosition;
+    private List<Row> peerRows = new ArrayList<>();
+    private boolean peersCollected = false;
+    private Row prevRow = null;
+
+    WindowBatchIterator(WindowDefinition windowDefinition, BatchIterator<Row> source, AggregateCollector collector) {
+        assert windowDefinition.partitions().size() == 0 : "Only empty OVER() is supported now";
+        assert windowDefinition.orderBy() == null : "Only empty OVER() is supported now";
+        assert windowDefinition.windowFrameDefinition() == null : "Only empty OVER() is supported now";
+
+        this.windowDefinition = windowDefinition;
+        this.source = source;
+        this.collector = collector;
+    }
+
+    private boolean arePeers(@Nullable Row prevRow, Row nextRow) {
+        // all rows are peers when orderBy is missing
+        return windowDefinition.orderBy() == null;
+    }
+
+    @Override
+    protected BatchIterator<Row> delegate() {
+        return source;
+    }
+
+    @Override
+    public Row currentElement() {
+        return currentElement;
+    }
+
+    @Override
+    public void moveToStart() {
+        super.moveToStart();
+        sourceRowsConsumed = 0;
+        rowPosition = 0;
+        peerRows = new ArrayList<>();
+    }
+
+    @Override
+    public boolean moveNext() {
+        if (peersCollected && rowPosition < sourceRowsConsumed) {
+            rowPosition++;
+            return true;
+        }
+
+        while (source.moveNext()) {
+            sourceRowsConsumed++;
+            Row currentRow = source.currentElement();
+            if (arePeers(prevRow, currentRow)) {
+                peersCollected = false;
+                peerRows.add(currentRow);
+                prevRow = currentRow;
+            } else {
+                collectPeers();
+                rowPosition++;
+                return true;
+            }
+        }
+
+        if (source.allLoaded()) {
+            if (!peersCollected) {
+                collectPeers();
+            }
+
+            if (rowPosition < sourceRowsConsumed) {
+                // we still need to emit rows
+                rowPosition++;
+                return true;
+            } else {
+                return false;
+            }
+        } else {
+            source.loadNextBatch().toCompletableFuture().join();
```

Comment:
I think we'll have to do this in a non-blocking way (returning false to let the consumer call loadNextBatch on *this* batchIterator, which then delegates to the source)